In [1]:
import pandas as pd
import numpy as np
from geopy.geocoders import Nominatim
import time

In [2]:
df = pd.read_csv('Baton_Rouge_Traffic_Crash_Incidents_20250920.csv')

In [3]:
df.head()

,INCIDENT NUMBER,CRASH DATE,STREET ADDRESS,CITY,STATE,INTERSECTING STREET,DISTRICT,ZONE,CRASH OCCURRED ON,TYPE OF ROADWAY,...,INTERSECTION,TOTAL VEHICLES,VEHICLE ALIGNMENT,RELATION TO ROADWAY,MANNER OF COLLISION,PRIMARY FACTOR,SECONDARY FACTOR,LATITUDE,LONGITUDE,GEOMETRY
0,25-195467,2025 Sep 19 01:00:00 AM,300 I-110,Baton Rouge,LA,CHIPPEWA ST,10,4G-1,Interstate,ONE-WAY ROAD,...,False,2,Straight,On roadway,Sideswipe - same direction,Violations,Movement prior to crash,30.478015,-91.167079,POINT (-91.1670849 30.4780204)
1,25-195210,2025 Sep 18 01:00:00 AM,11000 COURSEY BLVD,Baton Rouge,LA,ROYAL ASCOT,11,3C-3,City street,TWO-WAY ROAD WITH A PHYSICAL BARRIER,...,True,2,Straight,On roadway,Front to rear - rear end,Violations,Movement prior to crash,30.415450,-91.062569,POINT (-91.0625749 30.4154554)
2,25-195143,2025 Sep 18 01:00:00 AM,2400 S SHERWOOD FOREST BLVD,Baton Rouge,LA,W.I12,8,3E-2,City street,TWO-WAY ROAD WITH A PHYSICAL BARRIER,...,True,2,Straight,On roadway,Angle - left opposite direction,Violations,Movement prior to crash,30.429276,-91.056297,POINT (-91.0563029 30.4292815)
3,25-195207,2025 Sep 18 01:00:00 AM,4747 S SHERWOOD FOREST BLVD,Baton Rouge,LA,COURSEY,8,3F-3,City street,TWO-WAY ROAD WITH A PHYSICAL BARRIER,...,False,2,Straight,On roadway,Angle - right into flow,Violations,Movement prior to crash,30.410795,-91.051446,POINT (-91.0514517 30.4108006)
4,25-195096,2025 Sep 18 01:00:00 AM,2800 O'NEAL LN,Baton Rouge,LA,NaN,9,3F-1,City street,TWO-WAY ROAD WITH A PHYSICAL BARRIER,...,False,2,Straight,On roadway,Front to rear - rear end,Violations,Movement prior to crash,30.427169,-91.007492,POINT (-91.0074983 30.4271743)


In [4]:
df.describe()

,DISTRICT,INJURY,FATALITY,HIT AND RUN,PEDESTRIAN INVOLVED,TRAIN INVOLVED,TOTAL VEHICLES,LATITUDE,LONGITUDE
count,45723.000000,45723.000000,45723.000000,45723.000000,45723.00000,45723.000000,45723.000000,45723.000000,45723.000000
mean,8.151477,0.773506,0.002712,0.226035,0.01520,0.000591,2.001159,30.443450,-91.121031
std,2.713685,0.418567,0.052007,0.418267,0.12235,0.024294,0.452574,0.035057,0.048271
min,1.000000,0.000000,0.000000,0.000000,0.00000,0.000000,1.000000,30.341962,-91.210461
25%,6.000000,1.000000,0.000000,0.000000,0.00000,0.000000,2.000000,30.420603,-91.160619
50%,8.000000,1.000000,0.000000,0.000000,0.00000,0.000000,2.000000,30.438549,-91.132364
75%,11.000000,1.000000,0.000000,0.000000,0.00000,0.000000,2.000000,30.462766,-91.084638
max,12.000000,1.000000,1.000000,1.000000,1.00000,1.000000,9.000000,30.594465,-90.999871


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45723 entries, 0 to 45722
Data columns (total 29 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   INCIDENT NUMBER      45721 non-null  object 
 1   CRASH DATE           45723 non-null  object 
 2   STREET ADDRESS       45723 non-null  object 
 3   CITY                 45723 non-null  object 
 4   STATE                45723 non-null  object 
 5   INTERSECTING STREET  45548 non-null  object 
 6   DISTRICT             45723 non-null  int64  
 7   ZONE                 45723 non-null  object 
 8   CRASH OCCURRED ON    45723 non-null  object 
 9   TYPE OF ROADWAY      45723 non-null  object 
 10  ROAD SURFACE         45723 non-null  object 
 11  ROADWAY CONDITION    36 non-null     object 
 12  WEATHER CONDITION    45723 non-null  object 
 13  LIGHTING CONDITION   45723 non-null  object 
 14  INJURY               45723 non-null  int64  
 15  FATALITY             45723 non-null 

In [6]:
# Check if there is null records
df.isnull().sum()

INCIDENT NUMBER            2
CRASH DATE                 0
STREET ADDRESS             0
CITY                       0
STATE                      0
INTERSECTING STREET      175
DISTRICT                   0
ZONE                       0
CRASH OCCURRED ON          0
TYPE OF ROADWAY            0
ROAD SURFACE               0
ROADWAY CONDITION      45687
WEATHER CONDITION          0
LIGHTING CONDITION         0
INJURY                     0
FATALITY                   0
HIT AND RUN                0
PEDESTRIAN INVOLVED        0
TRAIN INVOLVED             0
INTERSECTION               0
TOTAL VEHICLES             0
VEHICLE ALIGNMENT          0
RELATION TO ROADWAY        0
MANNER OF COLLISION        0
PRIMARY FACTOR             0
SECONDARY FACTOR           0
LATITUDE                   0
LONGITUDE                  0
GEOMETRY                   0
dtype: int64

In [7]:
df.columns

Index(['INCIDENT NUMBER', 'CRASH DATE', 'STREET ADDRESS', 'CITY', 'STATE',
       'INTERSECTING STREET', 'DISTRICT', 'ZONE', 'CRASH OCCURRED ON',
       'TYPE OF ROADWAY', 'ROAD SURFACE', 'ROADWAY CONDITION',
       'WEATHER CONDITION', 'LIGHTING CONDITION', 'INJURY', 'FATALITY',
       'HIT AND RUN', 'PEDESTRIAN INVOLVED', 'TRAIN INVOLVED', 'INTERSECTION',
       'TOTAL VEHICLES', 'VEHICLE ALIGNMENT', 'RELATION TO ROADWAY',
       'MANNER OF COLLISION', 'PRIMARY FACTOR', 'SECONDARY FACTOR', 'LATITUDE',
       'LONGITUDE', 'GEOMETRY'],
      dtype='object')

#### Drop column Roadway condition [column-wise] and incident number [row-wise]

In [8]:
df.drop(['ROADWAY CONDITION'], inplace=True, axis=1)

In [9]:
df.dropna(subset=['INCIDENT NUMBER'], inplace=True)

#### Imputing column intersecting street

Getting the road name using the property latitude and longitude as there are 0 empty values for this. To retrieve the road name, a package geopy is used, and the values for intersecting street are populated for all 175 empty records. 

In [10]:
# first 5 rows where 'INTERSECTING STREET' is null 
df[df['INTERSECTING STREET'].isnull()].head()

,INCIDENT NUMBER,CRASH DATE,STREET ADDRESS,CITY,STATE,INTERSECTING STREET,DISTRICT,ZONE,CRASH OCCURRED ON,TYPE OF ROADWAY,...,INTERSECTION,TOTAL VEHICLES,VEHICLE ALIGNMENT,RELATION TO ROADWAY,MANNER OF COLLISION,PRIMARY FACTOR,SECONDARY FACTOR,LATITUDE,LONGITUDE,GEOMETRY
4,25-195096,2025 Sep 18 01:00:00 AM,2800 O'NEAL LN,Baton Rouge,LA,NaN,9,3F-1,City street,TWO-WAY ROAD WITH A PHYSICAL BARRIER,...,False,2,Straight,On roadway,Front to rear - rear end,Violations,Movement prior to crash,30.427169,-91.007492,POINT (-91.0074983 30.4271743)
21,25-194402,2025 Sep 16 01:00:00 AM,12900 AIRLINE HWY,Rural East Baton Rouge,LA,NaN,9,3F-3,US highway,TWO-WAY ROAD WITH A PHYSICAL BARRIER,...,False,2,Straight,On roadway,Angle - perpendicular/other angle,Violations,Movement prior to crash,30.376429,-91.034698,POINT (-91.034704 30.3764345)
81,25-193276,2025 Sep 13 01:00:00 AM,3348 DRUSILLA LN,Baton Rouge,LA,NaN,11,3G-4,Off road/private property,ONE-WAY ROAD,...,False,2,Straight,In parking lane or zone,Not a collision between two motor vehicles in ...,Violations,Movement prior to crash,30.419822,-91.088728,POINT (-91.0887338 30.4198275)
113,25-192790,2025 Sep 12 01:00:00 AM,I-110 HWY,Baton Rouge,LA,NaN,5,4E-1,Interstate,ONE-WAY ROAD,...,False,2,Straight,On roadway,Not a collision between two motor vehicles in ...,Movement prior to crash,Not applicable,30.496016,-91.161090,POINT (-91.1610959 30.496021)
286,25-190641,2025 Sep 07 01:00:00 AM,768 I-12 HWY,Baton Rouge,LA,NaN,8,3F-1,Interstate,TWO-WAY ROAD WITH A PHYSICAL BARRIER,...,False,2,Straight,On roadway,Sideswipe - same direction,Violations,Movement prior to crash,30.439151,-91.019868,POINT (-91.0198741 30.4391562)


In [11]:
geolocator = Nominatim(user_agent="study_crash_pattern")

In [12]:
def get_intersecting_street(row):
    try:
        location = geolocator.reverse(f"{row['LATITUDE']}, {row['LONGITUDE']}", exactly_one=True)
        address = location.raw.get('address', {})
        return address.get('road', 'Not Found') 
    except Exception as e:
        return 'Error'
    finally:
        time.sleep(1)

In [13]:
# Create a filter for missing value
missing_mask = df['INTERSECTING STREET'].isnull()

# Apply .loc accessor to select and update the respective value after applying the function get_intersecting_street
df.loc[missing_mask, 'INTERSECTING STREET'] = df[missing_mask].apply(get_intersecting_street, axis=1)

#### Check if there are any empty records

In [14]:
df.isnull().sum()

INCIDENT NUMBER        0
CRASH DATE             0
STREET ADDRESS         0
CITY                   0
STATE                  0
INTERSECTING STREET    0
DISTRICT               0
ZONE                   0
CRASH OCCURRED ON      0
TYPE OF ROADWAY        0
ROAD SURFACE           0
WEATHER CONDITION      0
LIGHTING CONDITION     0
INJURY                 0
FATALITY               0
HIT AND RUN            0
PEDESTRIAN INVOLVED    0
TRAIN INVOLVED         0
INTERSECTION           0
TOTAL VEHICLES         0
VEHICLE ALIGNMENT      0
RELATION TO ROADWAY    0
MANNER OF COLLISION    0
PRIMARY FACTOR         0
SECONDARY FACTOR       0
LATITUDE               0
LONGITUDE              0
GEOMETRY               0
dtype: int64

In [15]:
df.head()

,INCIDENT NUMBER,CRASH DATE,STREET ADDRESS,CITY,STATE,INTERSECTING STREET,DISTRICT,ZONE,CRASH OCCURRED ON,TYPE OF ROADWAY,...,INTERSECTION,TOTAL VEHICLES,VEHICLE ALIGNMENT,RELATION TO ROADWAY,MANNER OF COLLISION,PRIMARY FACTOR,SECONDARY FACTOR,LATITUDE,LONGITUDE,GEOMETRY
0,25-195467,2025 Sep 19 01:00:00 AM,300 I-110,Baton Rouge,LA,CHIPPEWA ST,10,4G-1,Interstate,ONE-WAY ROAD,...,False,2,Straight,On roadway,Sideswipe - same direction,Violations,Movement prior to crash,30.478015,-91.167079,POINT (-91.1670849 30.4780204)
1,25-195210,2025 Sep 18 01:00:00 AM,11000 COURSEY BLVD,Baton Rouge,LA,ROYAL ASCOT,11,3C-3,City street,TWO-WAY ROAD WITH A PHYSICAL BARRIER,...,True,2,Straight,On roadway,Front to rear - rear end,Violations,Movement prior to crash,30.415450,-91.062569,POINT (-91.0625749 30.4154554)
2,25-195143,2025 Sep 18 01:00:00 AM,2400 S SHERWOOD FOREST BLVD,Baton Rouge,LA,W.I12,8,3E-2,City street,TWO-WAY ROAD WITH A PHYSICAL BARRIER,...,True,2,Straight,On roadway,Angle - left opposite direction,Violations,Movement prior to crash,30.429276,-91.056297,POINT (-91.0563029 30.4292815)
3,25-195207,2025 Sep 18 01:00:00 AM,4747 S SHERWOOD FOREST BLVD,Baton Rouge,LA,COURSEY,8,3F-3,City street,TWO-WAY ROAD WITH A PHYSICAL BARRIER,...,False,2,Straight,On roadway,Angle - right into flow,Violations,Movement prior to crash,30.410795,-91.051446,POINT (-91.0514517 30.4108006)
4,25-195096,2025 Sep 18 01:00:00 AM,2800 O'NEAL LN,Baton Rouge,LA,O'Neal Lane,9,3F-1,City street,TWO-WAY ROAD WITH A PHYSICAL BARRIER,...,False,2,Straight,On roadway,Front to rear - rear end,Violations,Movement prior to crash,30.427169,-91.007492,POINT (-91.0074983 30.4271743)


In [16]:
df['CRASH DATE'] = pd.to_datetime(df['CRASH DATE'])

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 45721 entries, 0 to 45722
Data columns (total 28 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   INCIDENT NUMBER      45721 non-null  object        
 1   CRASH DATE           45721 non-null  datetime64[ns]
 2   STREET ADDRESS       45721 non-null  object        
 3   CITY                 45721 non-null  object        
 4   STATE                45721 non-null  object        
 5   INTERSECTING STREET  45721 non-null  object        
 6   DISTRICT             45721 non-null  int64         
 7   ZONE                 45721 non-null  object        
 8   CRASH OCCURRED ON    45721 non-null  object        
 9   TYPE OF ROADWAY      45721 non-null  object        
 10  ROAD SURFACE         45721 non-null  object        
 11  WEATHER CONDITION    45721 non-null  object        
 12  LIGHTING CONDITION   45721 non-null  object        
 13  INJURY               45721 non-null 

In [18]:
# Count duplicate incident number
duplicate_count = df['INCIDENT NUMBER'].duplicated().sum()
print(f"BEFORE: Found {duplicate_count} duplicate incident numbers.")

BEFORE: Found 336 duplicate incident numbers.


In [19]:
df.drop_duplicates(subset=['INCIDENT NUMBER'], keep='first', inplace=True)

In [20]:
duplicate_count = df['INCIDENT NUMBER'].duplicated().sum()
print(f"AFTER: Found {duplicate_count} duplicate incident numbers.")

AFTER: Found 0 duplicate incident numbers.


#### Updating weather condition values for simplicity

In [21]:
df['WEATHER CONDITION'].unique()

array(['Clear', 'Cloudy', 'Unknown', 'Rain', 'Other', 'Fog, smog, smoke',
       'Severe crosswinds', 'Freezing rain or freezing drizzle', 'Snow',
       'Blowing snow'], dtype=object)

In [22]:
# Convert all entries to lowercase and remove leading/trailing whitespace
df['WEATHER CONDITION'] = df['WEATHER CONDITION'].str.lower().str.strip()

In [23]:
# Define a mapping from the old, complex values to the new, clean ones
weather_mapping = {
    'fog, smog, smoke': 'fog',
    'freezing rain or freezing drizzle': 'freezing rain',
    'blowing snow': 'snow'
    # 'severe crosswinds' could be grouped into 'other' if you want fewer categories
}

# Apply the replacement to the column
df['WEATHER CONDITION'].replace(weather_mapping, inplace=True)

/var/folders/4z/9ddz1dxn051gf97hvy69q3f80000gn/T/ipykernel_79881/345774084.py:10: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['WEATHER CONDITION'].replace(weather_mapping, inplace=True)


In [24]:
df['WEATHER CONDITION'].unique()

array(['clear', 'cloudy', 'unknown', 'rain', 'other', 'fog',
       'severe crosswinds', 'freezing rain', 'snow'], dtype=object)

In [25]:
df['DAY_OF_WEEK'] = df['CRASH DATE'].dt.day_name()

In [26]:
df['CRASH_MONTH'] = df['CRASH DATE'].dt.month

#### Standarize categorical data such that all values are in same format

In [27]:
categorical_cols = df.select_dtypes(include=['object']).columns

In [28]:
for cat in categorical_cols:
    df[cat] = df[cat].str.lower().str.strip()

#### Encode all categorical values using one-hot encoding

In [29]:
# Get high cardinality columns
print(df.select_dtypes(include=['object']).nunique())

INCIDENT NUMBER        45385
STREET ADDRESS         17410
CITY                       2
STATE                      1
INTERSECTING STREET     4848
ZONE                      71
CRASH OCCURRED ON          6
TYPE OF ROADWAY            5
ROAD SURFACE              11
WEATHER CONDITION          9
LIGHTING CONDITION         8
VEHICLE ALIGNMENT          4
RELATION TO ROADWAY       11
MANNER OF COLLISION       22
PRIMARY FACTOR            12
SECONDARY FACTOR          13
GEOMETRY               42362
DAY_OF_WEEK                7
dtype: int64


In [30]:
# Drop columns that are identifier, have high cardinality values 

columns_drop = [
    'INCIDENT NUMBER', 
    'STREET ADDRESS', 
    'INTERSECTING STREET', 
    'GEOMETRY', 
    'STATE'
]

df_cleaned = df.drop(columns=columns_drop)

In [31]:
# Calculate the cumulative percentage contribution of each zone
zone_distribution = df_cleaned['ZONE'].value_counts(normalize=True).cumsum()

# Use .to_string() to ensure the full list prints out
# print(zone_distribution.to_string())

In [32]:
N = 40
top_n_zones = df_cleaned['ZONE'].value_counts().nlargest(N).index
df_cleaned['ZONE'] = df_cleaned['ZONE'].where(df_cleaned['ZONE'].isin(top_n_zones), 'Other')

In [ ]:
df_cleaned['CRASH DATE'] = df_cleaned['CRASH DATE'].dt.date

In [ ]:
original_columns = df_cleaned.columns
clean_cols = [col.lower().replace(' ', '_') for col in original_columns]
df_cleaned.columns = clean_cols

In [ ]:
# Grouping collision types
df_cleaned['manner_of_collision'] = df_cleaned['manner_of_collision'].str.lower()

collision_mapping = {
    # Rear-End
    'front to rear - rear end': 'Rear-End',
    
    # Sideswipe
    'sideswipe - same direction': 'Sideswipe',
    'sideswipe - opposite direction': 'Sideswipe',
    'sideswipe - left overtake': 'Sideswipe',
    'sideswipe - right overtake': 'Sideswipe',
    
    # Angle
    'angle - perpendicular/other angle': 'Angle',
    'angle - left opposite direction': 'Angle',
    'angle - left into flow': 'Angle',
    'angle - right into flow': 'Angle',
    'angle - left across flow': 'Angle',
    'angle - right across flow': 'Angle',
    'angle - left overtake': 'Angle',
    'angle - right overtake': 'Angle',

    # Backing
    'backing - rear to front': 'Backing',
    'backing - rear to side': 'Backing',
    'backing - rear to rear': 'Backing',
    
    # Head-On
    'front to front - head on': 'Head-On',
    'front to front - left against flow': 'Head-On',
    'front to front - right against flow': 'Head-On',
    
    # Single Vehicle
    'not a collision between two motor vehicles in transport': 'Single Vehicle / Non-Collision',
    
    # Other / Unknown
    'other': 'Other',
    'unknown': 'Unknown'
}

# Apply the replacement to the column
df_cleaned['manner_of_collision'] = df_cleaned['manner_of_collision'].replace(collision_mapping)

In [4]:
df_cleaned.to_csv('preprocessing-1.csv', index=False)

In [34]:

# 3. Now, perform one-hot encoding on the cleaned DataFrame
# This will encode all remaining 'object' type columns
# df_encoded = pd.get_dummies(df_cleaned, columns=df_cleaned.select_dtypes(include=['object']).columns)
